swap/src/seed.rs

```diff
@@ -18,6 +22,29 @@ impl Seed {
         Ok(Seed(bytes))
     }
 
+    /// Return the private key and chain code to be used as root extended
+    /// private key for a BIP32 wallet.
+    pub fn root_private_key(&self, network: bitcoin::Network) -> PrivateKey {
```

Comment:
We should use BDK here. Why are we calculating this manually?

Reply:
This is the right approach, it is just not fully complete:

1. We might want to create _sub-seed_ from where we derive the root private key, i.e.:
`[root-seed]` -> `[bitcoin-root-key-seed]` -> `b"Bitcoin seed"[bitcoin-root-key-seed]`
2. The return value of this function should be an `ExtendedPrivKey`.

An `ExtendedPrivKey` already does the last step of the derivation for you, see here: https://docs.rs/bitcoin/0.26.0/src/bitcoin/util/bip32.rs.html#491-510

As such, all you need to do is `[root-seed]` -> `[bitcoin-root-key-seed]` and then call `ExtendedPrivKey::new(network, [bitcoin-root-key-seed])`. (Pseudo-code, "array notation" is meant to be placeholder for binary data.)

Reply:
Can you clarify the relation with BDK? Do you derive manually and pass the result to BDK?

Reply:
BDK is descriptor-based wallet and as such is initialized with a descriptor. To support derivation, we need to give it a descriptor that supports derivation, probably BIP84 to be native-segwit.

A BIP84 descriptor still needs an extended private key (this patch calls it "root" key) that needs to come from somewhere. The only thing that is missing here is that we should be deriving an `ExtenedPrivKey` from our seed and not a single `PrivateKey`.

BDK will not tell you how to create an `ExtendedPrivKey`, it just requires it as a parameter. As such, the code that we see here cannot be offloaded to BDK because it is out-of-scope. It can be partly offloaded to `rust-bitcoin` (see the link in my last comment).

See the other comment for more context + links.